File: N-ary Tree Preorder Traversal/N-ary_Tree_Preorder_Traversal.cpp
// 和590多叉树的后序遍历可以看成是一类题目
// Runtime: 152 ms, faster than 88.63% of C++ online submissions for N-ary Tree Preorder Traversal.
// Memory Usage: 33.1 MB, less than 46.95% of C++ online submissions for N-ary Tree Preorder Traversal.

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution 
{
public:
    vector<int> preorder(Node* root) 
    {
        vector<int> res;

        // 边界条件处理
        if (root == nullptr)
            return res;

        stack<Node*> treeStack;
        treeStack.push(root);

        while (!treeStack.empty())
        {
            root = treeStack.top();
            treeStack.pop();

            for (int i = root->children.size() - 1; i >= 0; i--)
                treeStack.push(root->children[i]);
            
            res.push_back(root->val);
        }
        
        return res;
    }
};

File: N-ary Tree Preorder Traversal/N-ary_Tree_Preorder_Traversal.py
# Runtime: 100 ms, faster than 76.02% of Python3 online submissions for N-ary Tree Preorder Traversal.
# Memory Usage: 17.6 MB, less than 74.45% of Python3 online submissions for N-ary Tree Preorder Traversal.

"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        
        res = []

        if root is None:
            return res
        
        treeStack = []
        treeStack.append(root)

        while treeStack:

            root = treeStack.pop()

            for i in range(len(root.children) - 1, -1, -1):
                treeStack.append(root.children[i])
            
            res.append(root.val)

        return res